### PR TITLE
Gemma bug fixes - Approx GELU, Layernorms, Sqrt(hd)

### PR DESCRIPTION
**Pull request by danielhanchen**
Just a few more Gemma fixes :) Currently checking for more as well! All fixes are derived from Unsloth's 2.5x faster and 70% less VRAM Gemma finetuning script :) https://github.com/unslothai/unsloth

Related PR: https://github.com/huggingface/transformers/pull/29285, which showed RoPE must be done in float32 and not float16, causing positional encodings to lose accuracy. @ArthurZucker @younesbelkada

### 1. Approx Gelu and not Exact
Activation function according to https://twitter.com/danielhanchen/status/1763613620909580505 (waiting for confirmation) should be approximate gelu and not exact gelu. Ie `gelu` should actually be `gelu_pytorch_tanh` which calls `PytorchGELUTanh` and that calls `nn.functional.gelu(input, approximate="tanh")`. `gelu` calls `nn.functional.gelu(input, approximate="none")`
Approx_Gelu_MqgVNPUOIJJL0PByVo10d

### 2. Layernorm (w+1) should be done in float32
The layernorms must be upcasted to float32 and not bfloat16 or float16 halfway. It must be done unlike Llama’s RMS Layernorm which downcasted before multiplying by the weights. We must downcast at the end.
RMS_Layernorm_IQDKDDYqOynaWPbbJT71e

### 3. sqrt(3072)=55.4256 but bfloat16 is 55.5
Interestingly, Gemma multiplies the embeddings by sqrt(hidden_dim). However, there is a precision problem! Gemma uses jnp.sqrt(self.embed_dim) .astype(x.dtype) which means sqrt(3072) = 55.4256, but casting it to bfloat16 rounds it to 55.5. For Gemma 2b, sqrt(2048) = 45.2548, but casting it to bfloat16 makes it 45.25.
Embedding_i3Hbs2CnyVrNF-otch_tL

### CODE REVIEW

**Comment by danielhanchen** [comment]
Actually I just noticed this requires updating all of Gemma's models on the HF Model Hub: https://huggingface.co/google/gemma-7b/blob/main/config.json for eg `gelu` should be changed to `gelu_pytorch_tanh`.

File: src/transformers/models/gemma/modeling_gemma.py [hunk 1]
```diff
@@ -169,7 +169,16 @@ def __init__(self, config):
...
+        hidden_act = config.hidden_act
+        if hidden_act != "gelu_pytorch_tanh":
+            logger.warning_once(
+                "Gemma's activation function should be approximate GeLU and not exact GeLU.\n"\
+                "Please edit your model config to use `gelu_pytorch_tanh` and not `gelu`.\n"\
+                "For now, we shall use `gelu_pytorch_tanh` temporarily.\n"\
+                "See https://github.com/huggingface/transformers/pull/29402 for more details."
+            )
+            hidden_act = "gelu_pytorch_tanh"
+        self.act_fn = ACT2FN[hidden_act]
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/gemma/modeling_gemma.py:181]: So we need a `self.hidden_activation` set to None by default and if `None` warn that we will use the new approx else use what was give

[the remark this answers is not in the document]
Comment (danielhanchen) [reply src/transformers/models/gemma/modeling_gemma.py:181]: Oh ok good point! Sorry didn't work on this in the meantime - I found a few more issues, and will push them here tomorrow :)

File: src/transformers/models/gemma/modeling_gemma.py [hunk 2]
```diff
@@ -170,7 +173,16 @@ def __init__(self, config):
         self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
         self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
         self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=False)
-        self.act_fn = ACT2FN[config.hidden_act]
+        hidden_act = config.hidden_act
+        if hidden_act != "gelu_pytorch_tanh":
```
Comment (younesbelkada) [inline src/transformers/models/gemma/modeling_gemma.py:177]: This way even if there is a model with the old gelu on the config we're force-setting hidden_act to `"gelu_pytorch_tanh"` right? 
I think we should either use a new config name or create a new attribute in the config `force_use_exact_gelu`, that is iniailizaed to False so that users can have the flexibility to switch to the old act function in case they fine-tuned it with old GeLU, what do you think?

**Review by younesbelkada — COMMENTED** [review]
Thanks for working on this @danielhanchen ! 🤩 
I left one comment about backward compatibility, what do you think?

File: src/transformers/models/gemma/modeling_gemma.py (hunk 2 above)
Comment (danielhanchen) [reply src/transformers/models/gemma/modeling_gemma.py:177]: Hmm I like that approach ie `getattr(config, "force_use_exact_gelu", False)` so if `force_use_exact_gelu = True` then True. If `force_use_exact_gelu = False` then also False, and False otherwise.

Comment (younesbelkada) [reply src/transformers/models/gemma/modeling_gemma.py:177]: yes! I think we can add that directly into `GemmaConfig` class and default it to False

Comment (danielhanchen) [reply src/transformers/models/gemma/modeling_gemma.py]: I added `force_use_exact_gelu`!

File: src/transformers/models/gemma/modeling_gemma.py [hunk 3]
```diff
@@ -160,7 +163,7 @@ def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
     return q_embed, k_embed
 
 
-# Copied from transformers.models.mistral.modeling_mistral.MistralMLP with Mistral->Gemma
+# Edited from transformers.models.mistral.modeling_mistral.MistralMLP with Mistral->Gemma
```
Comment (ArthurZucker) [inline src/transformers/models/gemma/modeling_gemma.py:166]:
```suggestion
```
alright to remove entirely

**Review by ArthurZucker — APPROVED** [review]
LGTM ! Let's fix the CI and merge!

File: src/transformers/models/gemma/modeling_gemma.py [hunk 4]
```diff
@@ -894,7 +915,10 @@ def forward(
         hidden_states = inputs_embeds
 
         # normalized
-        hidden_states = hidden_states * (self.config.hidden_size**0.5)
+        # Gemma downcasts the below to float16, causing sqrt(3072)=55.4256 to become 55.5
+        # See https://github.com/huggingface/transformers/pull/29402
+        normalizer = torch.tensor(self.config.hidden_size**0.5, dtype = hidden_states.dtype)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply src/transformers/models/gemma/modeling_gemma.py]: Done!

**Comment by danielhanchen** [comment]
Closing this since we merged this! Thanks everyone!

**Comment by nxphi47** [comment]
@danielhanchen Many thanks for the fixes. Do you observe any performance difference before and after this fix? thanks.

File: src/transformers/models/gemma/modeling_gemma.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (johnzheng1975) [reply src/transformers/models/gemma/modeling_gemma.py]: Just one question: how to apply this in vllm configure?
